cardano-api/src/Cardano/Api/Block.hs

```diff
@@ -79,6 +91,14 @@ data Block era where
                   -> Consensus.ShelleyBlock (ShelleyLedgerEra era)
                   -> Block era
 
+-- | A block consists of a header and a body containing transactions.
+--
+pattern Block :: BlockHeader -> [Tx era] -> Block era
+pattern Block header txs <- (getBlockHeaderAndTxs -> (header, txs))
```

Comment:
We're attempting to hide away internal representations in `data Block era`, and present a simplified representation here: a block is just a block header and some transactions. Eventually we should probably not export the other constructor of `Block` except is some sort of `Internal` module.

An alternative to this is to keep `Block` abstract and provide projection functions i.e. export `getBlockHeaderAndTxs` or similar.

There are some advantages to using a pattern synonyms. It indicates to users that this is in some sense a "complete" pattern (GHC is smart enough to see this is complete). The pattern also appears in the same place as constructors in the haddocks. This is more inline with a "closed world" design meaning users can just rely on pattern matching rather than having to search for projection functions. We also seem to use pattern synonyms already in various parts of cardano-api.

@dcoutts, perhaps you want to chime in on this?

Reply:
I agree that in general view patterns are horrid, and we don't use them generally.

To use (non-trivial) pattern synonyms (which is sensible here) we're forced to use view patterns. It's a very limited use. It doesn't infect anything else.